### ISSUE TITLE
Support PHP 5

### ISSUE THREAD

**Issue by XedinUnknown** [php-fig/http-server-handler#3]
Hi,

I understand how it's EOL and all, but not everyone can afford to switch to PHP 7. For example, I want to use the standard in a WP plugin, and our company aims for PHP 5.4, because there are many users who are still on it. I believe that the interface should leave space for people to chose the implementation that they want, but also that it should not prevent older PHP users from implementing it. PHP 5.6 is still not EOL, and is another reason why the interface IMHO should support it.

The hint of return type for [`RequestHandlerInterface#handle()`] appears to be the only thing that makes this package require PHP 7. If it could be removed, and the PHPDoc standard could be used instead to declare the return type, then the minimal required version of PHP for this package could go as low as PHP 5.3. Is this a possibility?

[`RequestHandlerInterface#handle()`]: https://github.com/php-fig/http-server-handler/blob/master/src/RequestHandlerInterface.php#L17

**Comment by shadowhand** [comment]
As far as I am concerned, the answer to this should be definitely **no**. If you look at [PHP Supported Versions](http://php.net/supported-versions.php) security support for PHP 5.6 ends in 8 months, and PHP 7.0 in 7 months.

There is absolutely no valid reason to maintain PHP 5 support, and doing so is a security risk. FIG should not be complicit in allowing outdated PHP versions to continue to be used.

**Comment by XedinUnknown** [comment]
Still not understanding why FIG should **explicitly disallow** outdated (PHP 5.6 - not so outdated) versions to be used.

**Comment by XedinUnknown** [comment]
Also, IMHO it is the implementation's concern to be secure, and therefore to perhaps be targeted for PHP 7. A standard is made for interoperability, and having FIG standards which are not compatible with PHP 5 makes people develop non-interop implementations, which makes people more hesitant to switch to newer stuff, because their old stuff may break.

**Comment by shadowhand** [comment]
PHP 5.6 **is** outdated (no longer actively developed) and security support has only been extended because it is *the last version of PHP 5*.

> which makes people more hesitant to switch to newer stuff, because their old stuff may break

Testing can solve this for you. A tool like [php7cc](https://github.com/sstalle/php7cc) will help.

**Comment by XedinUnknown** [comment]
Moving to use newer things is not my problem, but is a problem for many people. I just have to deal with PHP 5, because the consumers of the software I make have a problem with newer things. IMHO, this is a valid reason to maintain PHP 5 support, at least where it can be done in a reasonably (excuse the pun) way. Here, it means simply removing the return type-hint, and allowing lower versions in Composer.

**Comment by Jean85** [comment]
Removing a return type is not free. For you, in PHP 5, it's meaningless, but I assure you that having such a hard constraint on the interface means that we are enforcing an hard requirement on the implementors.

It also has advantages with tools like static analysis and IDEs.

**Comment by XedinUnknown** [comment]
I understand the significance of the return type hint for PHP 7. However, being compliant with a standard seems again to be a concern of the implementation, rather than of the standard. Also, static code analysis tools such as Phan have no problem recognizing PHPDoc return type declarations.

**Comment by shadowhand** [comment]
Removing it eliminates runtime type safety and requires additional testing to be created.

**Comment by XedinUnknown** [comment]
Again, I completely understand that, and agree. On the other hand, a few hundred thousand environments would have a chance of using interop implementations.

**Comment by shadowhand** [comment]
Speaking purely for myself, I am fine with not having those environments be able to use official middleware if it means better runtime safety for everyone else.

**Comment by stefanotorresi** [comment]
@XedinUnknown please have a look at the first sentence in this page: https://www.php-fig.org/bylaws/mission-and-structure/

TL;DR: We want things to move *forward*, not have the maximum possible adoption rate.

**Comment by XedinUnknown** [comment]
If FIG is so adamant about promoting standards even at cost of adoption rate, how is it that one of the recently accepted standards, [PSR-11](https://packagist.org/packages/psr/container), supports PHP 5.3? Also, I understand that this may soon have its minimal requirements raised, but a version for PHP 5 will still exist. IMHO, this is the best course of action for all interface packages. Why not release a version of PSR-15 which supports PHP 5.3, and then carry on to 7 and above? This way, we can have our new standards for when we can afford to support them, and the adoption rate due to legacy support.

**Comment by Jean85** [comment]
PSR-11 didn't have any way of using a return or scalar type (or others PHP 7+ features), because a container can hold anything (`@return mixed`); so, the package in itself doesn't have anything that requires having an higher PHP version requirement.

Also, PSR-11 had to take care of BC with pre-existing DI packages, where a return type on the interface would make it incompatible with current implementations, forcing would-be implementors to release new major versions due to this breaking change.

OTOH, in this case the return type is significant, we do not have pre-existing implementors and the trade off was weighted and considered meaningful. 

Last but not least, I think your example has a fallacy: having two versions of the standard would just mean that no one would have any incentive in using the latter one, because it would just mean a lower adoption rate. We recognize that we are slightly forcing our hand doing this, but it's something that we desire and strive to voluntarily.

**Comment by stefanotorresi** [comment]
> If FIG is so adamant about promoting standards even at cost of adoption rate

No one is adamant about anything, it's a matter of trade-offs.
I empathise with your frustration, I really do, but please realise that developing a standard meant to be novel and long-lasting on something that is going EOL in less than a year makes no sense, and it doesn't make any sense either to have two standards of the same abstraction at the same time, because they would compete with each other and in this specific case we would end up with two middleware based ecosystems with a lot of implementation duplication, which is hardly desirable.

**Comment by XedinUnknown** [comment]
I appreciate your empathy, @stefanotorresi. But I do not see how there may be two eco-systems stemming from this. It's the same exact interface with the exact same FQN coming from the same package, with one being less strict _in terms of language_ (yet conceptually identical). If some implementations return something other than the declared return type, this means that they are not compatible, and cannot be seen as part of a separate yet valid eco-system.

With regard to having two "versions", how about the following PoV. If PSR-11 gets its minimal requirements raised, the old version will stay, and this will permit both PHP 5 and PHP 7 implementations to comply. Do you think this will have a negative impact on the Container eco-system? Does this warrant removal of the legacy version that supports PHP 5? I don't think so. The only consequence is that old code that was interop before will continue being interoperable. It will comply with a legacy version of the standard, and yet it will comply. Further, if the param/return types don't actually change, then implementations complying with legacy standard will also comply with the newer one. The same can be achieved by making a PHP 5 compatible release of the PSR-15 package, and any other package.

I understand that all I can do is ask, and I am the one that has to prove anything here, but could someone be so kind as to explain to be the negative effects of what I am suggesting?

**Comment by stefanotorresi** [comment]
> If PSR-11 gets its minimal requirements raised, the old version will stay

nope, the older version would be marked as deprecated, as happened with PSR-0 when it was superseded by PSR-4.

**Comment by XedinUnknown** [comment]
A new standard that supersedes a legacy one is not the same thing as having a new version of the same standard, though.

**Comment by stefanotorresi** [comment]
> A new standard that supersedes a legacy one is not the same thing as having a new version of the same standard, though.

Yes it is the same thing, because the new one will always meant to replace an older one. If it's not meant to do so, why would we want to have a new one in the first place?

You don't want multiple standards of the same subject, because it defeats the purpose of it being a standard! If you release a new one, the old one *must* be deprecated. This was NIH, we took inspiration from the IETF style, and although we have discussed alternatives in how to handle PSR versioning, nothing actionable has come out of it yet, so we stick to the original process:

- a new PSR on the exact same subject supersedes the old one.
- there shall be *one* PSR per subject.
- implementors are, for obvious reasons, strongly encouraged to not use deprecated PSRs.

These are not actual rules for the FIG, I'm just stating how things have been done so far as deemed appropriate, but maybe we should clarify such points on an official capacity.

**Comment by XedinUnknown** [comment]
No, it is not the same thing.

If `my/package` depends on `psr/simple-cache:^1.0`, and `psr/simple-cache:1.0.1` is released, then my code still works, still complies, and is interop with code that uses `psr/simple-cache:1.0.1`.

If `my/package` depends on `psr/container:^1.0`, and `psr/container-2:1.0` comes out, deprecating `psr/container`, then my code no longer works, complies, or is interoperable with code that uses `psr/container-2:1.0`.

By the way, thanks for clarifying those points. If I could vote, I would vote for those to be included somewhere if not as a rule, then as a guide for FIG members, and the consumers of FIG standards.

**Comment by stefanotorresi** [comment]
I think that's the crux you're missing: as stated before, we're not following semver with PSR numbering.

A PSR gets an incremental number and is a self contained document. Whether it is compatible with a previous PSR that it replaces or not is irrelevant: the new one is the one to be used, the old one is dead.

For example, despite PSR-4 being BC with PSR-0, the latter has been marked as deprecated regardless, because we don't want to foster any ambiguity whatsoever: if an implementor wants to use a standard, they should never have any more than one choice, because **there shall be one standard per subject, otherwise we failed**.

Composer packages are released under the SemVer scheme because that's how Composer and Packagist work, but we would never release anything different from typos or errata. Releasing something with a different platform requirement under the same name is completely out of question, for the time being.

Again, this is an outstanding issue, we realise that the IETF versioning scheme is somewhat limited, and in hindsight a different choice could have been made, but this is how it works for now:

### Updating platform requirements needs a new PSR to be made, and a new PSR on the same exact subject means that the old one gets deprecated.

**Comment by XedinUnknown** [comment]
> Updating platform requirements needs a new PSR to be made, and a new PSR on the same exact subject means that the old one gets deprecated.

Understood, thanks!

> there shall be one standard per subject, otherwise we failed.

How about PSR-6 and PSR-16?

I think the best course of action for me then is to fork PSR-15 packages, and make them PHP 5 compatible, but keep all the same exact FQNs. Is this OK? I would like to be as close with FIG as possible. I also aim to create helpful standards.

**Comment by stefanotorresi** [comment]
> How about PSR-6 and PSR-16?

These have not the *same exact subject*, in fact they have different names, different goals, different use cases, hence the contracts are different and they are hosted under different Composer packages, they can co-exist and one can be composed with the other. The rationale is explained more thoroughly in the [PSR-16 metadoc](https://www.php-fig.org/psr/psr-16/meta/).
On a personal note, I've always felt that PSR-16 also aimed to answer to the very cold reception PSR-6 had, because of some underlying design oversights that were (rightfully so imho) heavily criticised by some of the major potential implementors. 
Please appreciate also that these standards take years to make and each one has its own more or less complex history: PSR-6 was also the tipping point where it was acknowledged that the FIG needed a restructuring and a more convoluted development process (all which became the so called FIG 3.0 bylaw) to avoid repeating the same mistakes again.

**Comment by XedinUnknown** [comment]
Thank you very much for your explanation, @stefanotorresi. I appreciate and understand the effort that goes into PSRs, because myself and my colleague also create standards as part of @Dhii. We would like one day to become a complement to FIG, by providing a fast-paced environment for initiatives and experiments. I also read and understand the new bylaws, and I appreciate why they are that way.

With regard to PSR-16 and PSR-6, I can say that I completely see the point of both standards. One is great for memoization, and the other - for fined-grained generic cache control. They both reflect different aspects of caching, which perhaps could not be normalized at the time.

Still, could someone commend on my question about forking please?

**Comment by Jean85** [comment]
Still, to have them available in your current environment, you would be forced to fork (and fix) any implementor to be used.

I.e., implementors of this package are not PHP 5 friendly:
 * [`duncan3dc/sessions`](https://packagist.org/packages/duncan3dc/sessions): `^7.1`
 * [`objective-php/router`](https://packagist.org/packages/objective-php/router): `^7.0`
 * [`relay/relay`](https://packagist.org/packages/relay/relay): `^7.0`

And I've just took the top 3 with most stars, without looking inside; since they require PHP 7, their code will not run under PHP 5.

**Comment by XedinUnknown** [comment]
I think you misunderstood. I am the implementor, and I want to comply.

**Comment by Jean85** [comment]
Oh ok! Sorry, that's the reason why I couldn't understand what you were saying.

Well, in that case you could do it, but you either use a different FQCN (breaking interop) or you would use a different one, breaking other implementors if the end user application is using any of them.

**Comment by XedinUnknown** [comment]
> or you would use a different one

Sorry, didn't understand. Different what?

**Comment by Jean85** [comment]
Sorry, I wrote it wrong. I've edited the original message.

**Comment by XedinUnknown** [comment]
Thanks. How would using the same FQCN break interop? I think this is in fact the way to preserve it.

**Comment by Jean85** [comment]
Let's talk code!

 * you would be using `Psr\Http\Server\RequestHandlerInterface` for you fork
 * you would just remove the return type
 * you would implement your package with this fork

This would mean:
 * you package will work only with your forked `RequestHandlerInterface`
 * if an app uses the original `RequestHandlerInterface`, it will not be able to use your package (no real interop)

**Comment by XedinUnknown** [comment]
@Jean85, I was thinking that Composer's ]`replace`][[composer\replace] would solve the problem of clashing FQNs in interface packages. As for what the packages will work for, when declaring accepted or returned types, people would specify e.g. `Psr\Http\Server\RequestHandlerInterface`, and this would allow retrieving/accepting my implementation of the handler, as well as usage of that handler in the exact same way as the original - except for type safety. Looks like everything will still work. Am I wrong?

[composer\replace]: https://getcomposer.org/doc/04-schema.md#replace

**Comment by XedinUnknown** [comment]
@oscarotero, this would definitely save the day. Unfortunately, the namespace is different, making the FQN different, meaning that substitution is not possible.

**Comment by weierophinney** [comment]
> But I do not see how there may be two eco-systems stemming from this.

This one is easy: middleware written for the PHP 5 variant _would not be compatible_ with the PHP 7 variant, and would require _rewriting_ once the user was ready to move to PHP 7 and the PHP 7 version of the interfaces.

Why? The type hints. If the user updates to PHP 7 and then the PHP 7 version of the interfaces, their code will not fulfill the interfaces due to missing type hints. Since their code will still work with the PHP 5 version of the interfaces, they would most likely stick with those, as it represents _less work_ for them. Essentially, any update from the PHP 5 to PHP 7+ versions _requires a major version bump_ of your own code, as it represents _a change in API_.

This is not trivial, and not desired. The point of having the type hints, particularly the return type hints, is to provide a guarantee of consistency of operation. Since we have a version of PHP that allows usage of them, and because the only supported version that _does not_ support them is fast reaching its EOL for even security updates, it made sense to target a version that supports them.

This may mean that adoption is initially slow. However, standards are intended _to last a long time_. As a standards body, we're looking at the long tail of adoption. This is true for standards bodies such as IETF and WHAT-WG as well; they recognize that adoption takes time, and focus on getting details _correct_ first.

**Comment by oscarotero** [comment]
@XedinUnknown Maybe you can use aliases to use these older namespace with new names. This is what was done to migrate the older http-interop spec to new psr maintaining the BC.
https://github.com/http-interop/http-server-middleware/commit/1ed99649e5f0d785c16d53cc021d7187ec350f28

**Comment by weierophinney** [comment]
> Perhaps I'm missing something, but I do not see how this is true. Narrowing of the return type does not break LSP, which means that code which returns a subset of what it is allowed to return still works. But maybe here my knowledge of PHP 7 is lacking. Am I wrong?

In this case, the PHP 5 code would be _expanding_ the return type presented by the PHP 7 interfaces, because it would be _omitting_ the return type.

You can definitely narrow the return type. When PSR-15 was in the pre-PHP 7 http-interop packages, I was able to add a return type to my implementations without an issue. However, it doesn't go the other way around; if an interface provides a return type, you cannot omit it from an implementation. So if you started with http-interop and did not define a return type, when you adopted PSR-15, you couldn't just change the namespace of the interfaces; your code still broke because the return type was omitted from your implementation.
